docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,77 @@
+
+### P2P Transport
+The peer-to-peer transport layer is implemented using `libp2p`, which provides critical features such as NAT holepunching, node discovery, relaying, address determination, encrypted message transport, and more. 
+
+The most important feature provided by `libp2p` is baked in NAT holepunching support. While this is very well supported by the library, its important to understand the limitations that are naturally imposed by NATs, and the resulting contraints placed on `libp2p`. The following are some of the major challenges that have to be considered:
+
+* Hole-punching requires a relay node to perform the synchronization step of the DCUTR protocol. Both end nodes need to be able to connect to this temporary relay, so a dialing node must know beforehand what relays a destination node is attached to. Habitat identifies relays through community configuration, but this implies that each community must have at least one relay. In the future, it could be advantageous to optionally use external nodes as relays to perform synchronization.
+* `libp2p` automatically identifies potential relay nodes by having each node ask its peers to attempt to dial it across the public internet. If these peers are able to reach the node without a relay, and if the node presents a consisten IP address (its not behind a symetric NAT), then it can be marked as reachable and become a relay. Importantly, `libp2p` requires that four distinct peers make a successful connection before a node can be marked reachable. This might make things challenging for small communities. Even if they have a publicly accessible node, if they don't have enough peers for `libp2p` to confirm the node's public status, it will be unable to act as a relay, and therefore NAT holepunching capabilities will not be available to the community. 
+
+### Node Services
+The node layer comprises all facilities in Habitat that are shared between communities. These include:
+
+* Reverse proxy
+* Distributed file system
+* Process manager
+* Data sources
+
+#### Reverse Proxy
+An important part of the design philosophy of Habitat is to open as few ports as necessary. Since a Habitat instance could potentially run dozens of applications at at time, a reverse proxy is used to map requests to the proper application based on the path expressed in the URL.
+
+The reverse proxy is served on port 2041. In addition, the reverse proxy is available via `libp2p` transport on port 2042. The `libp2p` reverse proxy just redirects to the first.
+
+There are two main types of reverse proxy rules. Redirect rules redirect traffic to another server running on the nodee. In the future, this should also be able to redirect traffic to another node as well. The second is file server rules, which allow for a directory of files to be served, most likely for web content. 
```

Comment:
anything to add about constraint on reverse proxy rules? like are there any conflicts with Habitat paths and application paths or no?